proposals/0000-local-modules.rst

```diff
@@ -0,0 +1,1330 @@
+      called ``qual``) followed
+      by ``qualified``, the export
+      QEnv from the current module includes ``qual ↦ orig`` (where ``orig``
+      is the original name found by looking up ``qual`` in the ambient QEnv).
+
+      If ``qual`` has no qualifier, report a warning controlled by
+      ``-Wunqualified-qualified-exports``. In this case, the ``qualified``
+      keyword is a no-op.
+
+      If there are bundled identifiers exported (in ``(...)``), these are
+      also exported with the same qualification as the root identifier,
+      even if the bundled identifiers are written with different qualifications.
+      (Example: the export item ``M.T(K1, M.K2, N.K3) qualified`` exports
+      ``M.T``, ``M.K1``, ``M.K2``, and ``M.K3``.)
+
+      This behavior is different than the behavior without the ``qualified``
+      keyword, which causes the unqualified name to be mapped to the original
+      name in the export environment (as is done prior to this proposal).
+
+      In code::
+
+        interpretExpItem(haskell98item@(modids0.occname(..)) 'qualified', ambient_env)
+          = { modids0.occ1 ↦ orig | modids1.occ1 ∈ haskell98item
+                                  , modids1.occ1 ↦ orig ∈ ambient_env }
+
+      The key part here is that the return is a qualified-name mapping. Note
+      that the qualification in the exported mapping is ``modids0``, not the
+      ``modids1`` that might be found in the list of bundled qualified names.
+      See the ``M.T`` example a few paragraphs above.
+
+   #. For an export item like ``module qualified modids [ exports ]``:
+
+      If ``exports`` is given, we must interpret this export specification into
+      an export QEnv. Let ``E`` be the ambient QEnv. Then, ``exports`` is
+      interpreted with respect to the QEnv ``strip(modids, E)`` (that is, ``E``
+      with ``modids.`` stripped off from every domain element). Call this
+      intepretation of ``exports`` to be ``X'``. Let ``X = qualify(modids, X')``.
+
+      If ``exports`` is not given, then let ``X = qualify(modids, strip(modids, E))``. That is,
+      ``X`` includes all bindings in ``E``
+      whose domain element is qualified by ``modids``.
+
+      This export item exports the QEnv ``X``.
+
+      In this way, the export item ``module qualified M ( f, g )`` means the
+      same as ``module qualified M ( f ), module qualified M ( g )``.
+
+      In code::
+
+        interpretExpItem('module' 'qualified' modids expspec, ambient_env)
+          = let X' = interpretExpSpec(expspec, strip(modids, ambient_env)) in
+            qualify(modids, X')
+        interpretExpItem('module' 'qualified' modids, ambient_env)
+          = qualify(modids, strip(modids, ambient_env))
+
+   #. For an export item like ``module modids [ exports ]``:
+
+      The ambient, in-scope QEnv is ``E``. Define a new QEnv ``E'``
+      as follows::
+
+        E' = { qual ↦ orig | qual ↦ orig ∈ E
+                           , modids.qual ↦ orig ∈ E }
```

Comment:
Wouldn’t it be nice if Haskell had a syntactic way of distinguishing binding sites from occurrences in patterns, even if only to be able to use it in pseudo-code like this? ;-)

Reply:
Yes, indeed. :)